feat(switch): add `<rh-switch>`

zeroedin opened the pull request:
## What I did

1. Added `<rh-switch>`

Closes #770

### TODO

- [x] Update JSDoc
- [x] Add README.md
- [x] Add changeset
- [x] Update [overview, style, guidelines, accessibility] docs pages


## Testing Instructions

1. View [Deploy Preview demo](https://deploy-preview-1513--red-hat-design-system.netlify.app/elements/switch/demo/)

## Notes to Reviewers

bennypowers commented on .changeset/wild-steaks-play.md:
```diff
@@ -0,0 +1,5 @@
+"@rhds/elements": minor
+---
+
+✨ Added `<rh-switch>`
```
Example?

bennypowers commented on elements/rh-switch/demo/demo.css:
Remove

bennypowers commented on elements/rh-switch/demo/reversed.html:
```diff
@@ -0,0 +1,41 @@
+        <span data-state="on">Message when on</span>
+        <span data-state="off" hidden>Message when off</span>
+      </label>
+      <rh-switch id="reversed" reversed></rh-switch>
```
Is the reversed boolean attribute strictly necessary?

zeroedin replied:
There is no boolean attribute on `rh-switch`  so yes that needs removed.  

This demo was copied over from `pf-switch`, it doesn't exist there either, so should fix there as well.  Not even sure we need this demo at all.  The only thing it shows is that html order of label doesn't matter which is true with any form element as long as it uses a `for` attribute.  

I'll remove the attr for now but maybe we just delete the demo in full.

bennypowers reviewed (commented):
Preliminary review:

zeroedin commented:
will rebase this when https://github.com/RedHat-UX/red-hat-design-system/pull/1502 merges

zeroedin commented on elements/rh-switch/rh-switch.css:
```diff
@@ -0,0 +1,66 @@
+  margin-inline: var(--rh-space-sm, 6px);
+  font-size: var(--rh-font-size-body-text-xs, 0.75rem);
+  color: var(--rh-color-text-primary-on-dark, #ffffff);
+}
```
Just pushed code that adds this, PTAL.  Didn't realize there was a dark mode form inputs in Figma so missed that, thanks for the heads up.

marionnegp replied:
I adjusted the colors in dark theme for better contrast in [this branch](https://www.figma.com/file/Dy3AAEJadjosLTZIuBsQp0/chansey-updates?type=design&node-id=106-12044&mode=design) of the RHDS library file. 

- The label for the disabled state uses `var(--rh-color-text-secondary-on-dark, #c7c7c7)`
- The background fill for the switch in dark theme didn't pass color contrast, so I had to change it to `var(--rh-color-gray-40, #A3A3A3)`
- For consistency, the label for the light theme disabled switch now uses `var(--rh-color-text-secondary-on-light, #4d4d4d)`

Any objections to these changes, @coreyvickery?

zeroedin commented:
@marionnegp @coreyvickery 

I went ahead and implemented @marionnegp's suggestion above.  I also corrected the focus ring color for dark mode which we missed.

Just let me know if something else changes.

zeroedin commented:
During our accessibility meeting today we came up with a way forward for accessible labels.  Thank you @nikkimk for leading and organizing and ideation of the solution.  As well as @hellogreg for the on demand screen reader testing.  Below is a TODO list of follow up items:

- [ ] Fix styles for when the component is used with a label first or following the switch and its aria-describedby element.
- [ ] Update any documentation around the API
- [ ] Fix demos

bennypowers reviewed (changes requested):
Some minor technical points. Otherwise, and with the ux improvements, good

marionnegp commented:
@zeroedin @coreyvickery, what do you think of positioning the label above the switch and having the type style match that of our form's labels? ([Figma file](https://www.figma.com/file/4ZPa83Jta0kwyh9aTS5daS/rh-switch-with-label?type=design&node-id=8%3A2&mode=dev&t=bFO0lbp0Gjp87zRV-1))

coreyvickery commented:
@marionnegp Looks good to me.

bennypowers commented on docs/_plugins/rhds.cjs:
From a weird merge? Should we reset this?

zeroedin replied:
this comment still valid?  where you reviewing prior to me updating the branches?

zeroedin commented on elements/rh-switch/docs/10-style.md:
Kicked back to draft.  We are waiting on updated design spec and updated docs given the refactor on accessibility feed back dealing with state messages and `<label>`.

marionnegp commented on elements/rh-switch/rh-switch.css:
```diff
@@ -0,0 +1,122 @@
+}
+
+:host(:disabled) .dark {
+  --_switch-track-background-color: var(--rh-color-gray-60, #4d4d4d);
```
Should be `--rh-color-gray-70`

marionnegp reviewed (changes requested):
Some changes to the text color for the slotted message:
- [ ] When the switch is off, text should use `--rh-color-text-secondary-on-light` in light contexts and `--rh-color-text-secondary-on-dark` in dark.
- [ ] When switch is disabled, text should be `--rh-color-gray-50` in light contexts and `--rh-color-gray-40` in dark. (We'll have to add a disabled text token color...)
- [ ] Can the color context demo show disabled switches?

marionnegp commented:
@coreyvickery, @zeroedin saw that `<rh-switch>`'s track disappears against `--rh-color-surface-dark`. 
![Screenshot 2024-05-15 at 1 44 55 PM](https://github.com/RedHat-UX/red-hat-design-system/assets/95588923/d28f3e3f-a6b5-4f13-9e5d-e53faf842d3c)

What do you think about changing the track color to `#4D4D4D` (`--rh-color-gray-60`) when switch is used on `--rh-color-surface-dark`? Here's a comparison of the disabled and off states if we do this.
![Screenshot 2024-05-15 at 1 46 06 PM](https://github.com/RedHat-UX/red-hat-design-system/assets/95588923/4444db16-86e6-4d84-a3ed-3258d6b58671)

coreyvickery commented:
@marionnegp Looks good, proceed.

bennypowers commented:
is there any way to avoid making `aria-describedby` a required public attr?

bennypowers commented:
looks like there was a bad merge :/

zeroedin commented:
> is there any way to avoid making `aria-describedby` a required public attr?

I'm don't know of or I am not seeing a way that isn't going to be bound by cross root aria issues, given that the data messages are in lightdom slotted.  

In the future maybe using `#internals.ariaDescription` would be useful here?

@hellogreg @adamjohnson any additional thoughts on this?

bennypowers reviewed (changes requested):
I made some changes here:
- messages are usually plain text, which lends itself to attributes, so
- added two slot/attr pairs for on/off messages
- used internals (falling back to `aria-describedby` when unsupported) to set aria relationship with slots when needed
- redid the demos for brevity, e.g. the colour context demo should just be to show that it works, but pattern-level stuff should go elsewhere
in the case of sibling labels, i found that page css was unnecessary anyways, since labels inherit color from the surface.

I ran VO tests on firefox, chrome, and safari, and my changes don't appear to have made things any worse.

zeroedin commented on elements/rh-switch/rh-switch.ts:
```diff
@@ -0,0 +1,194 @@
+     * thats when the consumer requests an update.  Switching between lighter -> light for example will
+     * not trigger the component to update at this time.
+     */
+    this.#internalColorPalette = this.closest('[color-palette]')?.getAttribute('color-palette');
```
there is a need for specific color-palette because to make given a specific color way accessible https://github.com/RedHat-UX/red-hat-design-system/pull/1513#discussion_r1601747740

```
:host(:disabled) .color-palette-dark {
  --_switch-track-background-color: var(--rh-color-gray-60, #4d4d4d);
}
```

which uses the `.color-palette-dark`.

zeroedin replied:
related to: https://github.com/RedHat-UX/red-hat-design-system/pull/1395

bennypowers replied:
So why not make it that color for all dark contexts?

zeroedin replied:
@marionnegp thoughts on this?

bennypowers replied:
Alternatively, solve #1395 without too-too many breaking changes, and block this pr on it

zeroedin replied:
@marionnegp Hard coded the disabled values 7e7f15a220c90bb78c1ed1a542b5d20500a42ee0 as we were discussing in office hours.

Please take a look at the disabled switches on the color context demo and let me know if this passes design muster.

https://deploy-preview-1513--red-hat-design-system.netlify.app/elements/switch/demo/color-context/

zeroedin replied:
resolved in @marionnegp commit 5d97936

zeroedin commented on elements/rh-switch/rh-switch.ts:
```diff
@@ -0,0 +1,194 @@
+            <path d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z" />
+          </svg>
+        ` : ``}
+        <div id="track">
+          <div id="handle-container">
+            <div id="handle"></div>
+          </div>
+        </div>
+      </div>
```
I guess I have a different opinion on this.  I find PFE's to be harder to understand to maintain the sum of its parts.  Whereas this implementation there is a clear delineation of each part and easier to read css that goes along with it, including no crazy calcs just to do simple transition of an element.

bennypowers replied:
Each node we add increases the performance footprint. Switch in particular is likely to appear many times on the page

zeroedin replied:
I hear you and would love to see the real world performance numbers on the difference between the implementations, slightly more html (3 divs) vs more crusty css.  Maybe we can pit fight `pf-switch` vs `rh-switch` at a later date,  no need for a debate here though and I'd like to get switch closed, so whatever helps to do that.

zeroedin replied:
@bennypowers pushed the update for this PTAL

marionnegp commented on elements/rh-switch/rh-switch.css:
```diff
@@ -0,0 +1,147 @@
+}
+
+:host(:disabled) .dark {
+  --_switch-track-background-color: var(--rh-color-gray-60, #4d4d4d);
```
The text color for dark disabled switches with labels should be `--rh-color-gray-40`. Everything else looks great! 

@coreyvickery, I created [an issue](https://github.com/RedHat-UX/red-hat-design-system/issues/1710) to update switch colors in the Figma library.

zeroedin replied:
@marionnegp The value I see in figma for a disabled track on dark is: `#383838`
![Screenshot 2024-07-11 at 5 03 45 PM](https://github.com/RedHat-UX/red-hat-design-system/assets/13666/23d81712-61f9-4bba-99c6-38b3282a5556)

This isn't the color you are wanting to change in the issue you created is it?

Changing the dark disabled track background color to gray-40 would make the track the same color as when its not disabled:
![image](https://github.com/RedHat-UX/red-hat-design-system/assets/13666/c140e13a-94c5-432c-b023-9e6577e29d14)

The property @bennypowers and I were discussing above was the text disabled color which was different based on which variant of dark it was on.  `#A3A3A3` as its in figma does not work on `dark` but does on `darker` `darkest`.  Because in figma we are only representing switch on one dark variant it was not caught that the text value isn't accessible on the `dark` variant background.  

I had added a work around for this to apply a different text color, but @bennypowers suggestion was to use the value from `dark` across all instead which is: `#4d4d4d` and I believe is a better compromise then the work around.  

Let me know if you want to meet about this to clarify, if I am off base here.

marionnegp replied:
No, `#4d4d4d` for the track color is correct here. I have an [issue](https://github.com/RedHat-UX/red-hat-design-system/issues/1710) to fix the dark, disabled switch in Figma based on what we have in the repo.

I was thinking about changing the text color for dark, disabled to `var(--rh-color-gray-40, #A3A3A3)`. It has more contrast than the current text color of `#707070` but still looks more subtle than the text color for enabled switches. This should be good for all dark color palettes. What do you think? (Also, sorry I missed the messages in a previous thread about the track color!!)

marionnegp commented on elements/rh-switch/docs/20-guidelines.md:
```diff
@@ -0,0 +1,253 @@
+</uxdot-example>
+
+
+## Best practices
```
The images under "Best practices" should be centered, and the example containers should be the same height. (If this is due to a difference of dimensions after exporting, I can re-export these images!)

marionnegp reviewed (changes requested):
Found a couple things in the docs after I sent in the last review

hellogreg commented:
`<rh-switch>` gets the all-clear on keyboard operation and screen readers. They’re all a little different, but that’s just how they are!

- ios safari/vo focus: switch a, description, message when on, switch button, on
- ios safari/vo toggle: off

- android chrome/talkback focus: on, switch a, switch, message when on
- android chrome/talkback toggle: off

- mac safari/vo focus: switch a, message when on, on, switch
- mac safari/vo toggle: switch a, message when on, on, switch

- win chrome/jaws focus: switch a, switch, pressed on, message when on
- win chrome/jaws toggle: message when off, off

- win firefox/nvda focus: switch, editable, on, message when on
- win firefox/nvda toggle: off, message when off

- win edge/narrator focus: switch a, on, message when on
- win edge/narrator toggle: switch a, off, message when off

- linux firefox/orca focus: switch a, switch on, message when on, focus mode
- linux firefox/orca toggle: off

hellogreg reviewed (approved):
Let Greg Toggle Messages
